Json branch

jeffh1996 opened the pull request:
All data is now  formatted in JSON format in the TelemetryReporting_json.cpp file. Currently using a serial port to send the data. The data is sent in a byte array over the port. Created new classes to support this new format. New classes are named <Pre-existing file name>_json.

bill-luu commented:
Enoch has merge rights.

@jeffh1996 After Enoch's comments, you should run this through AStyle to fix the formatting/whitespacing.

@enochtsang Is changing the Ui to select how the data being set in the scope of this PR?

enochtsang commented:
@bill-luu  No it is not.

enochtsang commented on src/BusinessLayer/CommunicationService_json.cpp:
```diff
@@ -0,0 +1,37 @@
+#include "CommunicationService_json.h"
```
Follow naming conventions for files, everything is in `CapCase` already.

jeffh1996 commented:
Coding and file changes:
1. Merged some of the new classes back together (the Communication Services class and the Battery Data class). The class for sending data in JSON format remains as its own class (renamed InternetReporting class).
2. The second Test Application class has been removed. The original Test Application now creates a new instance of the InternetReporting class in addition to the the original TelemetryReporting class. 

Functionality changes: 
1. Both "reporting" classes are using the same serial port to send data.  (Possible problem?) If requested, I can modify this to add a second serial port in which to send the data. 
2. When a button is clicked, both types of data are sent at the same time (JSON format and serial format). 

Note: This has not been run through Astyle yet.

enochtsang commented:
Reopen when you're done your changes.

jeffh1996 commented:
Changed so that InternetReporting is not hooked up to the view anymore

enochtsang commented on src/BusinessLayer/InternetReporting.h:
```diff
@@ -0,0 +1,69 @@
+class QIODevice;
+class CommunicationService;
+
+class InternetReporting : public QObject
```
If you're going to create an InternetReporitng, you need to change the name of the other service accordingly as well. `SerialReporting`.

enochtsang commented on src/BusinessLayer/CommunicationService.cpp:
```diff
@@ -28,21 +29,20 @@ void CommunicationService::setPeripheralSerialPort()
 
+void CommunicationService::setPeripheralInternet()
+{
+    outputDevice_ = new QSerialPort();
```
You can't just rename one thing if you're starting to do other stuff. Keep in mind the big picture. Right now it looks like the internet stuff got added. After you're done working with any code base, it should look like it's always been like that.

jeffh1996 commented:
Update: Travis build is working at this point

enochtsang commented on src/BusinessLayer/CommunicationService.h:
```diff
@@ -13,16 +15,19 @@ class CommunicationService : public QObject
     View& view_;
     QSerialPort* outputDevice_;
-    I_CommPeripheral* outputPeripheral_;
+    SerialPortPeripheral* outputPeripheral_;
```
member name should be changed

enochtsang commented on src/BusinessLayer/CommunicationService.cpp:
```diff
@@ -25,14 +26,30 @@ void CommunicationService::setPeripheralSerialPort()
+    outputPeripheral_->sendSerialData(packet, packetLength);
+}
+
+void CommunicationService::sendInternetData(const QByteArray & /*parameter name here*/)
```
You can choose a reasonable parameter name.

enochtsang commented on src/BusinessLayer/CommunicationService.cpp:
```diff
@@ -25,14 +26,30 @@ void CommunicationService::setPeripheralSerialPort()
+void CommunicationService::setInternetConnection()
 {
-    outputPeripheral_->sendData(packet, packetLength);
+    //function stub
```
General practice is to leave a TODO. This way you can search a codebase for just TODOs.

enochtsang commented on src/BusinessLayer/InternetReporting.cpp:
```diff
@@ -0,0 +1,494 @@
+QJsonObject Battery;
+QJsonArray CMUArray;
+QJsonArray mPPTArray;
+QJsonObject lightsInfo;
```
This data really doesn't need to shared as global variables. There are very very very few uses cases, that you need a global variable. Also in general, CapCase is reserved for types, if a word doesn't define a type, it shouldn't be CapCase.

Global variables are:
1. Confusing to read because they're hard to look for
2. Not testable
3. Not extensible

A much better alternative to your make functions is to return the created object rather than use a shared global variable.

enochtsang commented on src/BusinessLayer/InternetReporting.h:
```diff
@@ -0,0 +1,69 @@
+    void sendCmu();
+    void sendMppt();
+    void sendLights();
+    void sendAll();
```
We don't ever send this data separately, so we'll never have a parser that handles them separately. The only use case you need to handle is sending all the data at once.

jeffh1996 commented:
**Changes:**
1. Member variable names handling serial and Internet data are changed to be more distinguishable in the Communication service class.
2. Replaced function stubs with TODOs in the Communication service and the Internet peripheral classes.
3. Removed individual sending functions in the Internet reporting class. There is now only one sending function.
4. Removed global variables in the Internet reporting class. Helper functions now return the object. 
5. Renamed some variables in the Internet reporting class to proper formatting (No more CapCase).
6. Added a TODO in the Internet reporting class constructor.
7. Replaced parameters placeholders with parameter names (Added a Q_UNUSED macro because the compiler does not like unused parameters).

JordanHeinrichs commented:
Hey, Enoch asked me to take a look so here are my comments. Overall great work.

One design issue I noticed is I_CommPeripheral is now useless (it is also now not an interface but an abstract as functions are not pure virtual). The original point of it was to abstract what device the communication service is using. But that abstraction is now shattered as there is both sendSerialData and sendInternetData function within it.

This also means that SerialPortPeripheral has an empty sendInternetData function which is not correct.

As clearly that original abstraction is no longer wanted the best thing to do is simply delete I_CommPeripheral. I don't see any real uses of it as CommunicationService just uses both child classes directly. It should be easy and quick to remove.

That is the only major thing I noticed. A few white space issues and magic numbers are also present (InternetReporting::makeCmu for the magic numbers) but overall it seems good. 👍

jeffh1996 commented:
Alright. I will try and work on those issues. Thank you for your feedback.

jeffh1996 commented:
**Changes**
1. The I_CommPeripheral interface/abstract is now removed and all dependencies have been removed from the CommunicationService, InternetPeripheral, and SerialPeripheral classes/headers. 
2. More formatting on InternetReporting
3. Fixed the magic numbers in InternetReporting::makeCmu

Notes: I assume the magic numbers were the "i < 8" and "i < 15" in the InternetReporting::makeCmu function. I also noticed this in the SerialReporting::sendCmu function as well.

enochtsang commented:
Fix the merge conflicts and merge yourself.

jeffh1996 commented:
Did not realize those extra files were there. Deleted them

bill-luu commented:
Please put the includes in alphabetical order where possible, and once that's done run it through Astyle one more time.
